pkg/apis/core/v1beta1/helper/errors.go

```diff
@@ -63,25 +72,30 @@ var (
 	retryableConfigurationProblemRegexp = regexp.MustCompile(`(?i)(is misconfigured and requires zero voluntary evictions|SDK.CanNotResolveEndpoint|The requested configuration is currently not supported)`)
 )
 
-// DetermineError determines the Garden error code for the given error and creates a new error with the given message.
-// TODO(timebertt): this is should be improved: clean up the usages to not pass the error twice (once as an error and
-// once as a string) and properly wrap the given error instead of creating a new one from the given error message,
-// so we can use errors.As up the call stack.
-func DetermineError(err error, message string) error {
+// DetermineError determines the Gardener error codes for the given error and returns an ErrorWithCodes with the error and codes.
+func DetermineError(err error) error {
 	if err == nil {
-		return errors.New(message)
-	}
-
-	errMsg := message
-	if errMsg == "" {
-		errMsg = err.Error()
+		return nil
 	}
 
 	codes := DetermineErrorCodes(err)
 	if codes == nil {
-		return errors.New(errMsg)
+		return err
 	}
-	return &ErrorWithCodes{errMsg, codes}
+
+	// if error itself is ErrorWithCodes
+	if errorWithCodes, ok := err.(*ErrorWithCodes); ok {
+		err = errorWithCodes.ExtractError()
+		return &ErrorWithCodes{err, codes}
+	}
+
+	// if error contains ErrorWithCodes wrapped inside it
+	unwrappedError := errors.Unwrap(err)
+	if errorWithCodes, ok := unwrappedError.(*ErrorWithCodes); ok {
+		err = errorWithCodes.ExtractError()
+	}
```

Comment:
This is already covered by `DetermineErrorCodes`. I think, this can simply be dropped.